lib/src/git.rs

```diff
@@ -497,6 +497,47 @@ pub fn fetch(
     callbacks: RemoteCallbacks<'_>,
     git_settings: &GitSettings,
 ) -> Result<Option<String>, GitFetchError> {
+    // In non-colocated repositories, it's possible that `jj branch forget` was run
+    // at some point and no `jj git export` happened since.
+    //
+    // This would mean that remote-tracking branches, forgotten in the jj repo,
+    // still exist in the git repo. If the branches didn't move on the remote, and
+    // we fetched them, jj would think that they are unmodified and wouldn't
+    // resurrect them.
+    //
+    // Export will delete the remote-tracking branches in the git repo, so it's
+    // possible to fetch them again.
+    //
+    // For more details, see the `test_branch_forget_fetched_branch` test, and PRs
+    // #1714 and #1771
+    //
+    // Apart from `jj branch forget`, jj doesn't provide commands to manipulate
+    // remote-tracking branches, and local git branches don't affect fetch
+    // behaviors. So, it's unnecessary to export anything else.
+    //
+    // TODO: Create a command the user can use to reset jj's
+    // branch state to the git repo's state. In this case, `jj branch forget`
+    // doesn't work as it tries to delete the latter. One possible name is `jj
+    // git import --reset BRANCH`.
+    // TODO: Once the command described above exists, it should be mentioned in `jj
+    // help branch forget`.
+    let nonempty_branches = mut_repo
+        .base_repo()
+        .view()
+        .branches()
+        .iter()
+        .filter_map(|(branch, target)| target.local_target.as_ref().map(|_| branch.to_owned()))
+        .collect_vec();
+    // TODO: Inform the user if the export failed? In most cases, export is not
+    // essential for fetch to work.
+    let _ = export_some_refs(mut_repo, git_repo, |ref_name| {
+        matches!(
+            ref_name,
+            RefName::RemoteBranch { branch, remote:_ }
+               if !nonempty_branches.contains(branch)
+        )
```

Comment:
Can we apply the same filter passed to `import_some_refs()` (i.e. `common_filter_fn(ref_name) && !nonempty`)?

I don't think unrelated remote branches should be affected.

Reply:
I didn't do it our of laziness (since it doesn't matter much IMO), but I agree that it leaves a smell; I'll look into doing it properly.

Reply:
I added a commit doing this. It's a bit messy, let me know if you have ideas for making it cleaner. It's really annoying that you can't write:

```rust
let closure = if condition {
  |a| a > 0
} else {
  |_| true
};
```

Reply:
You can box the closure to dynamically dispatch them:
```rust
let closure: Box<dyn Fn(..) -> ..> = if condition {
    Box::new(|..| ..)
} else {
...
```

FWIW, it might be better for `git::fetch()` to always filter out non-remote branches. It doesn't make sense that `jj git fetch --branch *` doesn't import local branches, but `jj git fetch` does.

Reply:
> You can box the closure to dynamically dispatch them

I tried that, but I ran into trouble: where I was passing `*closure` to `import_some_refs`, Rust complained that it's not `Sized` but should be.

I abandoned the commit, maybe I can find what I did exactly.

> FWIW, it might be better for git::fetch() to always filter out non-remote branches. It doesn't make sense that jj git fetch --branch * doesn't import local branches, but jj git fetch does.

That's probably better, yes. I can add yet another commit to change that.

Reply:
> That's probably better, yes. I can add yet another commit to change that.

Hmm... I tried https://github.com/ilyagr/jj/commit/tmp-fetch-nolocalimport, and it's cleaner, but it fails a test:

```
failures:
    test_fetch_prune_deleted_ref

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 40 filtered out; finished in 0.24s


--- STDERR:              jujutsu-lib::test_git test_fetch_prune_deleted_ref ---
thread 'test_fetch_prune_deleted_ref' panicked at 'assertion failed: tx.mut_repo().get_branch(\"main\").is_some()', lib/tests/test_git.rs:1602:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/22e9fe644ea710eec50cb0aabcae7fa8dd9fd675/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/22e9fe644ea710eec50cb0aabcae7fa8dd9fd675/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/22e9fe644ea710eec50cb0aabcae7fa8dd9fd675/library/core/src/panicking.rs:117:5
   3: test_git::test_fetch_prune_deleted_ref
             at ./tests/test_git.rs:1602:5
   4: test_git::test_fetch_prune_deleted_ref::{{closure}}
             at ./tests/test_git.rs:1584:35
   5: core::ops::function::FnOnce::call_once
             at /rustc/22e9fe644ea710eec50cb0aabcae7fa8dd9fd675/library/core/src/ops/function.rs:250:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/22e9fe644ea710eec50cb0aabcae7fa8dd9fd675/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I don't have the energy now to figure it out; unless it's something really trivial, it can wait for another PR.

Reply:
> I tried that, but I ran into trouble: where I was passing `*closure` to `import_some_refs`, Rust complained that it's not `Sized` but should be.

Maybe you want `closure`, `&closure`, `&*closure`, or `|x| closure(x)`. `*closure` is of `dyn Fn..` type, which isn't sized.

> I don't have the energy now to figure it out; unless it's something really trivial, it can wait for another PR.

Sure. It's a separate thing we can insert before/after this PR.

Reply:
Nope, they all `impl Fn(..)..`. Anyway, we can clean it up later to unify `branch_name_globs == []` behavior.